changes/2024-09-13_cache-across-hierarchical-keyrings/background.md

```diff
@@ -0,0 +1,415 @@
+  i.e., currently, we do NOT support a shared cache
+  between multiple Hierarchical Keyrings.
+- Multiple CachingCMMs can use a single shared cache
+  (this is for native language implementations only,
+  there is, currently, no CachingCMM implemented
+  in Dafny)
+- We do NOT support a shared cache between multiple
+  Searchable Encryption Configurations.
+
+## Goals
+
+- We must allow caching across Key Stores/KMS Clients/KMS Keys
+  for multiple Hierarchical Keyrings and CachingCMMs.
+  For this, we need to strategically update the cache identifiers
+  for all materials ([background](#background)) in the CMC.
+- We MUST prevent Confused Deputy cases, i.e: the wrong Branch Key
+  Material being served
+- Our solution MUST be Easy to Use, Hard to Misuse
+
+## Out of Scope
+
+- Supporting a shared Cache in the DB-ESDK to fetch
+  BeaconKeyMaterials across multiple Searchable
+  Encryption Configurations, or across multiple KMS Relationships.
+
+## Issues and Alternatives:
+
+Before looking at the issues, please take a quick look at
+[Hierarchical Keyring Cache Identifier formula](#hierarchical-keyring-cache-identifier-formula),
+[CachingCMM Cache Identifier formula](#cachingcmm-cache-identifier-formula), and
+[DB-ESDK Searchable Encryption Cache Identifier formula](#db-esdk-searchable-encryption-cache-identifier-formula) for more context.
+
+Note that there are currently two hashings taking place.
+For example in the Hierarchical Keyring Cache Identifier formula,
+one hash is for the branch-key-id and the other is hashing
+the entire appended string
+(a part of which is the hash of the branch-key-id)
+to get the final cache identifier.
+We'll call these internal and external hashes.
+
+_Preferred options are in italics._
+
+### Issue 1: How to update the cache identifier for the Hierarchical Keyring to allow a shared cache?
+
+Note that according to the current
+[Hierarchical Keyring Cache Identifier formula](#hierarchical-keyring-cache-identifier-formula),
+if two KMS Relationships Hierarchical Keyrings are
+called with two different Key Stores but the same branch-key-id,
+there will be a collision (the branch-key-version will also
+have to co-incide in case of Decryption Branch Key Materials).
+
+#### _Option 1 (Recommended): Adding a Resource ID, Scope ID, Partition ID, and Resource Suffix (recommended)_
+
+We establish the following definitions for the
+Cache Entry Identifier formula:
+
+- **Resource Identifier:**
+  A Hex value that indicates
+  if an element is from a Caching CMM, Hierarchical Keyring,
+  or some other future resource.
+
+  - Caching_CMM : `0x01` (0001)
+  - Hierarchical_Keyring : `0x02` (0010)
+
+- **Scope Identifier:**
+  A Hex value that indicates
+  if an element is used for Encryption, Decryption,
+  Searchable Encryption, or some other future
+  purpose.
+
+  - Encrypt : `0x01` (0001)
+  - Decrypt : `0x02` (0010)
+  - Searchable Encryption : `0x03` (0011)
+
+- **Partition ID:**
```

Comment:
let's flip this around and talk about the default behavior first

Reply:
Sounds good. Thanks for the suggestion.